### ISSUE TITLE
Flatpak stops working normally when apparmor.d is installed

### ISSUE THREAD

**Issue by IPlayZed** [roddhjav/apparmor.d#314]
Distro: Arch Linux
Kernel: 6.8.7-hardened1
Systemd: 255.4-2
Flatpak: flatpak 1:1.15.8-1
Apparmor version: apparmor 3.1.7-1
Apparmor.d version: apparmor.d-git 0.2202-1

There are multiple issues:

- Using a Flatpak distribution terminal emulator, like BlackBox will cause starting a Flatpak application from the terminal to fail with `error: linkat: No such file or directory`.
- Using a Flatpak distribution terminal emulator, like BlackBox will cause trying to update or install fail for user Flatpaks, for instance trying to install Edge:
```
Warning: While pulling runtime/org.mozilla.firefox.Locale/x86_64/stable from remote flathub: Unable to write detached metadata: linkat: No such file or directory
Error: While pulling app/org.mozilla.firefox/x86_64/stable from remote flathub: Unable to write detached metadata: linkat: No such file or directory
error: Failed to install org.mozilla.firefox: While pulling app/org.mozilla.firefox/x86_64/stable from remote flathub: Unable to write detached metadata: linkat: No such file or directory
```
- Using a Flatpak distribution terminal emulator, like BlackBox will cause trying to view manpages to fail:
```
/usr/bin/man: can't execute less: Operation not permitted
/usr/bin/man: command exited with status 255: (cd /usr/share/man && /usr/lib/man-db/zsoelim) | (cd /usr/share/man && /usr/lib/man-db/manconv -f UTF-8:ISO-8859-1 -t UTF-8//IGNORE) | (cd /usr/share/man && preconv -e UTF-8) | (cd /usr/share/man && tbl) | (cd /usr/share/man && nroff -mandoc -rLL=191n -rLT=191n -Tutf8)
```
- Brave opens a window, but it is invisible.
- Firefox complains that it can not load or create the default profile.
- And probably many others I have not realized yet....

In the journal besides audit rotating the logs continually, there is not sign of anything.

I attached the audit log from this session, but there doesn't seem to be any denied action. See: [aa-log.log](https://github.com/roddhjav/apparmor.d/files/15055188/aa-log.log)

I also attached the ouput of `aa-status`, but Flatpak profiles does not seem to be in enforce mode, which is expected.... So I have no idea what goes wrong. See: [aa-status.log](https://github.com/roddhjav/apparmor.d/files/15055189/aa-status.log)

I am using systemd-homed managed users to operate my system. I have not tried if this comes out in the case of normal users.

For now, I am uninstalling apparmor.d, but if you get back to me in the issue I would be happy to provide you further info.

I am not sure if these problems are expected

**Comment by roddhjav** [comment]
Thanks for these logs. I am going to investigate this issue. Usually Flatpak apps works fine (but they need a bit more testing).

Can you try to get more logs with:
1. `aa-log -s -R` that should provide all apparmor log since boot time (if journalctl collect them)
2. `aa-log -R -f <nb>` where `<nb>` is `1`, `2`, `3` and `4` (the rotated audit log file)

> I am using systemd-homed managed users to operate my system. I have not tried if this comes out in the case of normal users.

That should not be a problem.

**Comment by roddhjav** [comment]
I have fixed a few issue with Firefox & Brave. All classic GUI app should work as expected. 

However, there are some limitations: Terminal & IDE

Both IDE and terminal need (indirect) full access to the system. For example, for terminal, it means, having `@{bin}/@{shells} rUx,`. It allows to start an unconfined shell from the profile so that you can run any command from your daily terminal, including a system update, a `rm -rf /`... There is a way to [restrict this further](https://apparmor.pujol.io/full-system-policy/), however, it includes more advanced configuration and security posture (ie: not something you expect when you are just installing some default profiles as in `apt install apparmor.d`)

The problem with flatpak is that the apparmor profile used is the same (`flatpak-app`) regardless of the application. So adding  `@{bin}/@{shells} rUx,` in it would allow every flatpak app to escape the profile... making it useless.

Possible solutions are:
- To not run your terminal under flatpak
- [To not confine flatpak app](https://github.com/roddhjav/apparmor.d/blob/e4c3f1f0769debe4081c8e1f890c66d97037ef99/apparmor.d/profiles-a-f/flatpak-app#L5-L18) (or to use an unconfined profile).

**Comment by IPlayZed** [comment]
Hi, Alex thanks for the fast reply. I was not around my PC, so I could not answer.

> Can you try to get more logs with:
> aa-log -R -s that should provide all apparmor log since boot time (if journalctl collect them)
> aa-log -R -f <nb> where <nb> is 1, 2, 3 and 4 (the rotated audit log file)

Do you still need these aforementioned logs?

> Both IDE and terminal need (indirect) full access to the system. For example, for terminal, it means, having @{bin}/@{shells} rUx,. It allows to start an unconfined shell from the profile so that you can run any command from your daily terminal, including a system update, a rm -rf /...

So if I understand correctly, all Flatpak terminal emulators without a profile applied to Flatpaks where this is restricted, can launch a shell which has unrestricted access? 

What I do not get is that according to the output of `aa-status` I have uploaded all Flatpak related profiles were in complain mode only. So why did it cause the `linkat: not found` problems?

> Possible solutions are:
> 
> To not run your terminal under flatpak
> [To not confine flatpak app](https://github.com/roddhjav/apparmor.d/blob/e4c3f1f0769debe4081c8e1f890c66d97037ef99/apparmor.d/profiles-a-f/flatpak-app#L5-L18) (or to use an unconfined profile).

As I do wish to maintain a more stable base system, if possible I would like to use the Flatpak version of my terminal emulator. Is there some way I can apply the currently set Flatpak profile to it _and_ apply the needed permissions to it so the shell launched in it can have all the access it needs to operate normally?

**Comment by roddhjav** [comment]
> Do you still need these aforementioned logs?

Yes, please, it could give some hint regarding linkat.

> So if I understand correctly, all Flatpak terminal emulators without a profile applied to Flatpaks where this is restricted, can launch a shell which has unrestricted access? Is this shell subject to restrictions set by bwrap to the emulator?

Yes, it is just that flatpak configure bwrap restriction according to the app developer setting (see for [BlackBox](https://github.com/flathub/com.raggesilver.BlackBox/blob/master/com.raggesilver.BlackBox.json). Meaning, there is no restriction for it. Please note that due to the nature of the sandboxed software, this is totally normal. Otherwise it would prevent you to do anything with your terminal...

> So why did it cause the linkat: not found problems?

In short, even in complain mode, apparmor can enforce some rules. See https://apparmor.pujol.io/issues/

> so the shell launched in it can have all the access it needs to operate normally?

Yes, apparmor.d is quite easy to personalise from a system admin side. Create the file `/etc/apparmor.d/localflatpak-app` with the following content, then restart apparmor and it should work:
```
  # The shell is not confined on purpose.
  @{bin}/@{shells} rUx,
```

**Comment by IPlayZed** [comment]
> Yes, apparmor.d is quite easy to personalise from a system admin side. Create the file /etc/apparmor.d/local/flatpak-app with the following content, then restart apparmor and it should work

Sorry for asking, but I am not that knowledgeable about creating Apparmor profiles: how do I identify the _exact_ Flatpak application I want to create the profile for? Let's say BlackBox is installed as a system-wide Flatpak. Do I have to specify it in the apparmor profile is some way?

The best I could find on this topic was this recommendation:

**/etc/apparmor.d/profile-name**
```
#include <tunables/global>

/profile/path/to/bwrap ix,
/profile/path/to/bwrap (argv) "/usr/bin/bwrap --args-for-your-app" cx
```

**Comment by roddhjav** [comment]
There is only one unique profile (`flatpak-app`) that is the same regardless of the flatpak application. You can extend it as I mentioned, however, it will apply these changes to all flatpak applications.

> The best I could find on this topic was this recommendation:

I do not know where you find this, but this is not a valid profile at all.

**Comment by IPlayZed** [comment]
> I do not know where you find this, but this is not a valid profile at all.

As proven again, LLMs are very creative regarding generating bogus 😄 

> There is only one unique profile (flatpak-app) that is the same regardless of the flatpak application. You can extend it as I mentioned, however, it will apply these changes to all flatpak applications.

So if I want my Flatpak terminal emulator to access a shell unrestricted, all Flatpaks would be allowed as well, if I add:
```
  # The shell is not confined on purpose.
  @{bin}/@{shells} rUx,
```

At this point I wonder if I should make a feature request for this for the upstream project, I think it would make sense to be able to specify per sandboxed applications, I mean Snap can do it, right?

**Comment by roddhjav** [comment]
Flatpak will not add this feature, because the security of flatpak sandbox does not come from apparmor/selinux but from bwrap. This is a (perfectly correct) technical choice they made long time ago, they won't change this. Snap supports it because some part of the sandbox security depends on apparmor (again, technical choice they made long time ago).

Outside of defense in depth (that could work in some condition), the real reason why I added a flatpak profile is because the purpose of this project is to ensure no program run unconfined.

**Comment by IPlayZed** [comment]
I see, thank you for enlightening me I am not nearly as well-versed as you are in this topic 😃 !
I will reinstall apparmor.d and get back to you with additional logs (also checking if they work fine otherwise).
Btw. would subsequent updates fail or overwrite my modified flatpak profile? I assume both the AUR package and the file I modify will write the same profile?

**Comment by IPlayZed** [comment]
@roddhjav I have installed `apparmor.d-git` from the AUR, created the following profile:
**/etc/apparmor.d/local/flatpak-app**:
```
@{bin}/@{shells} rUx,
```

However when running `flatpak run <your.flatpak.here>`, it still fails with the linkat error.
Did I create the profile incorrectly?

**Comment by IPlayZed** [comment]
> @roddhjav I have installed `apparmor.d-git` from the AUR, created the following profile: **/etc/apparmor.d/local/flatpak-app**:
> 
> ```
> @{bin}/@{shells} rUx,
> ```
> 
> However when running `flatpak run <your.flatpak.here>`, it still fails with the linkat error. Did I create the profile incorrectly?
> 
> BTW, there is no log for the `flatpak-app profile`. Here are the logs: [audit.log](https://github.com/roddhjav/apparmor.d/files/15241459/audit.log)
> 
> Again, no **DENIED** action at all 😒

Ah, so a bit of experimentation revelead when this problem occurs:

My default shell is bash for the user, however for BlackBox the custom startup command was set to be `fish`. I did this because I like fish better than bash, but I want to keep my configuration set in `.bashrc`.

It seems like if I delete this startup command, no such problem occurs. However if I launch fish specifically (launching bash in terminal doesn't act like it) manually or as a startup option, ther problem persists.

I suppose that fish is not registered in `@{shells}`?

**Comment by IPlayZed** [comment]
> I suppose that `fish` is not registered in `@{shells}`?

Ok a bit more update: it looks like this custom rule is ignored completely: commenting out the rule still produces the same behavior: from `bash` flatpaks work, from `bash` spawned `fish` shells, `linkat` is missing.

I guess, the problem is not with the `flatpak-app` profile... Which baffles me if there is no denied action. At the very least I would expect to see a denied action. (Or possibly the path to `linkat` is implicitly not available?)

**Comment by roddhjav** [comment]
Turns out the shell is not executed by the sandbox, but by `flatpak-session-helper` (which make sense). It does changes a few things. You don't need the additional rule (`@{bin}/@{shells} rUx,`) anymore, and terminal should work fine now.

Have a look at the commit comment in https://github.com/roddhjav/apparmor.d/commit/7963a65a88a034372c274894a4f179b7298ee477

**Comment by IPlayZed** [comment]
I tested it, it works :)

Thank you for the effort!
